src/main/java/com/networknt/schema/AnyOfValidator.java

```diff
@@ -24,40 +24,79 @@
+                                       parentSchema.getCurrentUri(),
+                                       schemaNode.get(i),
+                                       parentSchema));
+        }
+
+        if (config.isOpenAPI3StyleDiscriminators()) {
+            this.discriminatorContext = new ValidationContext.DiscriminatorContext();
+        } else {
+            this.discriminatorContext = null;
         }
     }
 
     public Set<ValidationMessage> validate(JsonNode node, JsonNode rootNode, String at) {
         debug(logger, node, rootNode, at);
 
+        if (config.isOpenAPI3StyleDiscriminators()) {
+            validationContext.enterDiscriminatorContext(this.discriminatorContext, at);
+        }
+
         Set<ValidationMessage> allErrors = new LinkedHashSet<ValidationMessage>();
         String typeValidatorName = "anyOf/type";
 
-        for (JsonSchema schema : schemas) {
-            if (schema.getValidators().containsKey(typeValidatorName)) {
-                TypeValidator typeValidator = ((TypeValidator) schema.getValidators().get(typeValidatorName));
-                //If schema has type validator and node type doesn't match with schemaType then ignore it
-                //For union type, it is must to call TypeValidator
-                if (typeValidator.getSchemaType() != JsonType.UNION && !typeValidator.equalsToSchemaType(node)) {
-                    allErrors.add(buildValidationMessage(at, typeValidator.getSchemaType().toString()));
-                    continue;
+        try {
+            for (JsonSchema schema : schemas) {
+                if (schema.getValidators().containsKey(typeValidatorName)) {
+                    TypeValidator typeValidator = ((TypeValidator) schema.getValidators().get(typeValidatorName));
+                    //If schema has type validator and node type doesn't match with schemaType then ignore it
+                    //For union type, it is must to call TypeValidator
+                    if (typeValidator.getSchemaType() != JsonType.UNION && !typeValidator.equalsToSchemaType(node)) {
+                        allErrors.add(buildValidationMessage(at, typeValidator.getSchemaType().toString()));
+                        continue;
+                    }
                 }
+                Set<ValidationMessage> errors = schema.validate(node, rootNode, at);
+                if (errors.isEmpty() && !config.isOpenAPI3StyleDiscriminators()) {
+                    return errors;
+                } else if (config.isOpenAPI3StyleDiscriminators()) {
+                    if (discriminatorContext.isDiscriminatorMatchFound()) {
+                        if (!errors.isEmpty()) {
+                            errors.add(buildValidationMessage(at, DISCRIMINATOR_REMARK));
+                        }
+                        return errors;
+                    }
+                }
+                allErrors.addAll(errors);
+//                allErrors.add(renderMissingMatchValidationMessage(at));

```

Comment:
@stevehu I wonder whether we can comment this line in - see the log renderer at the end of the file. At the moment failed `anyOf` validation doesn't give you an indication that no schema matched

Reply:
We have an issue opened to add more context information to the error messages to ensure they are understandable. If I am correct, the commented line will add an extra error message with more detailed info. Do you think it would be better to change the `errors.add(buildValidationMessage(at, DISCRIMINATOR_REMARK));` with the context info?

Reply:
that code line will specifically log an additional line saying that none of the candidates matched. Richer context is certainly nice and might make my commented line obsolete. The question just is when that will come and whether then we could simply remove the line again.

Btw I've commented the line as it has impact on test conditions on bunch of tests

Reply:
OK. Let's leave it this way for now. We need to resolve this issue in one shot for all validators.